datalad_next/tree.py

```diff
@@ -0,0 +1,1172 @@
+    """Return a function to exclude ``_TreeNode`` objects from the tree
+    (prevents them from being yielded by the node generator).
+
+    Returns
+    -------
+    Callable
+        Function that takes the Path object of a ``_TreeNode`` as input,
+        and returns true if the node should *not* be displayed in the tree.
+    """
+
+    def is_excluded(node: _TreeNode):
+        return any((
+            isinstance(node, FileNode) if not include_files else False,
+            node.path.name.startswith(".") if not include_hidden else False
+        ))
+
+    return is_excluded
+
+
+def increment_node_count(node_generator_func):
+    """Decorator for incrementing the node count whenever a ``_TreeNode`` is
+    yielded.
+
+    Parameters
+    ----------
+    node_generator_func: Callable
+        Function that yields ``_TreeNode`` objects
+    """
+    @wraps(node_generator_func)
+    def _wrapper(*args, **kwargs):
+        self = args[0]   # 'self' is a Tree instance
+        for node in node_generator_func(*args, **kwargs):
+            node_type = node.__class__.__name__
+            if node_type not in self.node_count:
+                raise ValueError(
+                    f"No counts collected for unknown node type '{node_type}'"
+                )
+            if node.depth > 0:  # do not count the root directory
+                # TODO: do not count symlinks if they point to
+                #  files/directories that are already included in the tree
+                #  (to prevent double counting)? Note that UNIX 'tree' does
+                #  count double.
+                self.node_count[node_type] += 1
+
+            yield node  # yield what the generator yielded
+
+    return _wrapper
+
+
+def yield_with_last_item(generator):
+    """Takes a generator and yields for each item, the item itself and
+    whether it is the last item in the sequence.
+
+    Returns
+    -------
+    Tuple[bool, Any]
+        A tuple (is_last_item, item)
+    """
+    prev_val = next(generator, None)
+    if prev_val is not None:
+        for current_val in generator:
+            yield False, prev_val
+            prev_val = current_val
+        yield True, prev_val
+
+
+def path_depth(path: Path, root: Path):
+    """Calculate directory depth of a path relative to the given root.
+
+    Can also be a negative integer if the path is a parent of the
+    tree root.
+
+    Returns
+    -------
+    int
+        Number of levels of the given path *below* the root (positive
+        integer) or *above* the tree root (negative integer)
+    """
+    if is_path_relative_to(path, root):
+        return len(path.relative_to(root).parts)
+    elif is_path_relative_to(root, path):
+        return - len(root.relative_to(path).parts)
+    else:
+        raise ValueError("Could not calculate directory depth: "
+                         f"'{path}' is not relative to the tree root "
+                         f"'{root}' (or vice-versa)")
```

Comment:
Given the frequency with which this function is called, I made an attempt to make it faster. The use of `is_path_relative_to()` prior to `relative_to()` essentially implies that `relative_to()` is ran at least twice on the same path combo.

Here is my alternative implementation:

```suggestion
    sign = 1
    try:
        rpath = path.relative_to(root)
    except ValueError:
        try:
            rpath = root.relative_to(path)
            sign = -1
        except ValueError:
            raise ValueError(
                "Could not calculate directory depth: "
                f"'{path}' is not relative to the tree root "
                f"'{root}' (or vice-versa)")
    return sign * len(rpath.parts)
```

Impact estimate (via `datalad -f disabled --on-failure ignore tree <path>`):

```
# on my 23k-dir /usr with no datasets (with patch)
            Mean        Std.Dev.    Min         Median      Max
real        5.187       0.457       4.253       5.184       5.932       
# on my 23k-dir /usr with no datasets (without patch)
real        5.851       0.851       4.388       5.763       7.191       

# on dataset with 80k directories across two levels (with patch)
real        8.709       0.995       7.370       9.005       9.752       
# on dataset with 80k directories across two levels (without patch)
real        9.380       0.559       8.608       9.621       9.911       
```

Not breathtaking, but 5-10% faster.

Reply:
Thank you for the proposed change! Looks much cleaner as well.

Indeed, `relative_to()` and its seemingly cheap `PurePath` string manipulation was eating up a huge chunk of runtime!

Got a significant speedup after cutting it out from several places (a551e25):

```shell
# dataset with 689 non-installed subdatasets

# before:
➜ python -m timeit -s 'from datalad_next.tree import TreeCommand; path="datasets.datalad.org/openneuro"' 'TreeCommand.__call__(path, dataset_depth=0, result_renderer="disabled")'
1 loop, best of 5: 7.37 sec per loop

# after:
➜ python -m timeit -s 'from datalad_next.tree import TreeCommand; path="datasets.datalad.org/openneuro"' 'TreeCommand.__call__(path, dataset_depth=0, result_renderer="disabled")'
1 loop, best of 5: 1.04 sec per loop
```

Reply:
Thank you for optimizing it further!

> Not sure what the use case could be for having datasets as subdirectories that are not submodules. Do you know if this is a common setup for any workflow?

`datalad-next` is providing similari-ish use case itself. The `datalad-annex::` git remote helper uses utility repositories underneath `.git` to function. Given that, I am not confident that there is no valid use case.

I will look into the `--dataset-depth` use case too.

Reply:
> datalad-next is providing similari-ish use case itself. The datalad-annex:: git remote helper uses utility repositories underneath .git to function. Given that, I am not confident that there is no valid use case.

That would already violate the current assumption that we need not look for datasets under the `.git` folder:

https://github.com/datalad/datalad-next/blob/070953dba2f7478c17c922a2d530e38512a6af48/datalad_next/tree.py#L853-L856

If these utility repos are purely used for internal machinery and are not generally of interest to the user, I would not mind skipping them. They could still be discovered in the 'full traversal' mode.

From a brief look at `datalad_annex.py` it seems like the utility repos are plain git(-annex) repos though, is that correct? In that case they would not be counted as datasets, since `tree` uses a narrow definition of dataset (only repos with a `.dataset/config` file -- even more accurate, but more expensive, would be to check for dataset ID). This is admittedly inconsistent with (most?) other datalad commands, which work on pure git repos too, and in general with the philosophy that a dataset is a repo. Thinking about it, we could provide an option `--repo-type=(datalad-only|all)` (default `datalad-only`) to detect all repos as datasets.

Reply:
OK, first look into `--dataset-depth` using your recommended testcase `///` with `-r -R2`. Indeed, no speed-up from limiting the traversal:

```
% datalad -f disabled tree --dataset-depth 1
datalad -f disabled tree --dataset-depth 1  20.16s user 8.46s system 103% cpu 27.538 total
% datalad -f disabled tree
datalad -f disabled tree  18.86s user 6.91s system 102% cpu 25.152 total
```

(First call only lists 33 datasets, second 2k+).

Reply:
I have not yet fully grokked the thing with the utility tree used to discover the "next" dataset. At least for the main/outer tree, I am fairly confident that I saw it behaving in an optimal fashion. Not touching paths that should not be touched.

For the inner tree (in `_generate_datasets()`)  I am still uncertain what it should be doing. I have a hunch that it needs to become aware of the `dataset_depth` parameter in some way to become efficient. But I guess I now understand your statement above:

> And why do we need to traverse the whole tree? Because (super)datasets could technically be anywhere. I wonder if we could make the assumption, though, that no top-level datasets are located underneath other datasets (as subdatasets that are not registered in gitmodules). Not sure what the use case could be for having datasets as subdirectories that are not submodules.

Which brings me to ask the question "what is the use case?" I am OK with "look for any top-level dataset anywhere" to not be faster than "show me everything", because I think this is not a use case where people would not be OK to wait.

You wrote:

> For example, displaying just the 33 first-level subdatasets of /// (installed with -r -R 2, without data) takes 1m20s! And that should be the hello world of command usage, not an edge case ;)

True, but in this case there is a single top-level dataset, and `datalad  subdatasets -d mydl/ -r -R1` provides the desired result in 200ms. We have that command, and people can use it. Your implementation makes no assumption of a single root dataset. It need not compete with the simple case.

I would propose two things to progress here:

1. It would be good to add some hints on performance expectations to `*Dataset discovery*`. Pretty much what you wrote above (it needs to traverse everything, because it wants to do more than a `subdatasets` call). So people can use the faster one, if that is enough for them.
2. Given the observation that a call with `--dataset-depth` is not faster than one without, there may be a chance to reduce the complexity of the implementation by simply always producing the full tree and only filter the Node outcome at the very end, rather than nesting the `Tree` generators. But that is just a hunch.

Apart from that I made a few small changes to make logging a bit faster (by prevent string formating when logging is not actually active (minuscule impact, but still the right thing to do).

```diff
diff --git a/datalad_next/tree.py b/datalad_next/tree.py
index ad16251..92a95aa 100644
--- a/datalad_next/tree.py
+++ b/datalad_next/tree.py
@@ -540,7 +540,7 @@ def is_dataset(path: Path, installed_only=False):
             # ignore symlinks even if pointing to datasets, otherwise we may
             # get duplicate counts of datasets
             lgr.debug("Path is a symlink, will not check if it points to a "
-                      f"dataset: '{path}'")
+                      "dataset: %r", path)
             return False
 
         if (path / ".datalad" / "config").is_file():
@@ -748,15 +748,16 @@ class Tree:
                     current_node.is_recursive_symlink(self.max_depth):
                 # if symlink points to directory that we may visit or may
                 # have visited already, do not recurse into it
-                lgr.debug(f"Symlink is potentially recursive, "
-                          f"will not traverse target directory: '{dir_path}'")
+                lgr.debug("Symlink is potentially recursive, "
+                          "will not traverse target directory: %r", dir_path)
                 return
 
             if current_node.exception is not None:
                 # if some exception occurred when instantiating the node
                 # (missing permissions etc), do not recurse into directory
                 lgr.debug("Node has exception, will not traverse directory: "
-                          f"path={current_node.path}, exc={current_node.exception}")
+                          "path=%r, exc=%r",
+                          current_node.path, current_node.exception)
                 return
 
             # sort child nodes alphabetically
@@ -893,9 +894,9 @@ class DatasetTree(Tree):
 
             except Exception as ex:
                 CapturedException(ex, level=10)  # DEBUG level
-                lgr.debug(f"Excluding node from tree because "
+                lgr.debug("Excluding node from tree because "
                           "an exception occurred while applying the "
-                          f"exclusion filter: '{node.path}'")
+                          "exclusion filter: %r", node.path)
                 return True  # exclude by default
 
             return False  # do not exclude
@@ -914,10 +915,7 @@ class DatasetTree(Tree):
         """Go to the next dataset or parent of dataset"""
         self._next_ds = next(self._ds_generator, None)
         if self._next_ds is not None:
-            lgr.debug(
-                f"Next dataset" +
-                (" parent" if isinstance(self._next_ds, DirectoryNode) else "")
-                + f": {self._next_ds.path}")
+            lgr.debug("Next dataset (parent): %s", self._next_ds.path)
 
     def _generate_datasets(self):
         """Generator of dataset nodes and their parent directories starting
```

When I get the next chance, I plan to look into computing some of the more expensive Node properties lazily. I have no good concept of a potential impact yet, but even little pieces shaved off of the Node instantiation time will have a solid cumulative impact.

Reply:
Thinking it over, I agree with your points.

> 1. It would be good to add some hints on performance expectations to *Dataset discovery*. Pretty much what you wrote above (it needs to traverse everything, because it wants to do more than a subdatasets call). So people can use the faster one, if that is enough for them.

Makes sense, since performance will hit a ceiling no matter the implementation. Done in b3d1ac3.

> 2. Given the observation that a call with --dataset-depth is not faster than one without, there may be a chance to reduce the complexity of the implementation by simply always producing the full tree and only filter the Node outcome at the very end, rather than nesting the Tree generators. But that is just a hunch.

I think it was a good hunch! For perceived command speed, it does not make that much of a difference IMO if the dataset tree is computed piecewise 'during' the execution of the outer main tree, or in its entirety before starting the main tree generator. Finding all datasets upfront also allows us to set a max depth for the main tree, based on the deepest found dataset (plus ``max_depth`` dataset subnodes).

The change is done in 79732a7. Since this now implies a potentially long wait right after running the command, we may want to consider hooking in a progress bar that monitors the dataset search. Do you think that would help or is it overkill?

Reply:
> Apart from that I made a few small changes to make logging a bit faster (by prevent string formating when logging is not actually active (minuscule impact, but still the right thing to do).

Happy to be proven wrong, but I was under the impression that f-strings outperformed all other string formatting constructs:

```shell
➜ python -m timeit -n 1000000 -s 'wi="wi"' '"sand%sch" % wi'
1000000 loops, best of 5: 116 nsec per loop

➜ python -m timeit -n 1000000 -s 'wi="wi"' '"sand{}ch".format(wi)'
1000000 loops, best of 5: 179 nsec per loop

➜ python -m timeit -n 1000000 -s 'wi="wi"' 'f"sand{wi}ch"'
1000000 loops, best of 5: 68 nsec per loop
```

Reply:
f-strings are the fastest interpolation, indeed! But what my patch does is to do *no interpolation at all* -- which is even faster ;-)

Essentially, unless a certain debug level is actually active, no CPU cycles will be wasted on string formatting/interpolation. `logging` will apply `%` formatting only at the point when an emitted log actually hits a formatters and is not excluded by any filter (where the level could just be one of many filters).

Reply:
Oh, my bad! You are passing the to-be-injected values as arguments to the logger, not to the `%` operator. So it's not a formatted string at all, until (if) the logger decides to evaluate it. That is so cool, definitely the right thing! (Off to rescuing CPU cycles I have been wasting en masse now...)

Reply:
;-) Thx!